src/psyclone/psyir/backend/fortran.py

```diff
@@ -0,0 +1,443 @@
+        elif isinstance(index, int):
+            # literal constant
+            dims.append(str(index))
+        elif index is None:
+            dims.append(":")
+        else:
+            raise NotImplementedError(
+                "unsupported get_dims index '{0}'".format(str(index)))
+    return dims
+
+
+def get_kind(symbol):
+    '''Infer the expected Fortran kind value from the Symbol
+    instance. This is a temporary LFRic-specific hack which simply
+    adds a hardcoded kind value for real variables and a hardcoded
+    kind value for integer variables. To work correctly in general the
+    symbol table needs some additional information added to it, see
+    issue #375.
+
+    :param symbol: The symbol instance.
+    :type symbol: :py:class:`psyclone.psyGen.Symbol`
+
+    :returns: the Fortran kind value for the symbol instance in lower \
+    case, or None if no kind value is required.
+    :rtype: str or NoneType
+
+    '''
+    kind = None
+    if symbol.datatype == "real":
+        kind = "r_def"
+    elif symbol.datatype == "integer":
+        kind = "i_def"
+    return kind
+
+
+class FortranPSyIRVisitor(PSyIRVisitor):
+    '''Implements a PSyIR-to-Fortran back end for PSyIR kernel code (not
+    currently PSyIR algorithm code which has its own gen method for
+    generating Fortran).
+
+    '''
+    def get_declaration(self, symbol):
+        '''Create and return the Fortran declaration for this Symbol.
+
+        :param symbol: The symbol instance.
+        :type symbol: :py:class:`psyclone.psyGen.Symbol`
+        :returns: The Fortran declaration as a string.
+        :rtype: str
+
+        '''
+        intent = get_intent(symbol)
+        dims = get_dims(symbol)
+        kind = get_kind(symbol)
+        result = "{0}{1}".format(self._nindent, symbol.datatype)
+        if kind:
+            result += "({0})".format(kind)
+        if dims:
+            result += ", dimension({0})".format(",".join(dims))
+        if intent:
+            result += ", intent({0})".format(intent)
+        if symbol.is_constant:
+            result += ", parameter"
+        result += " :: {0}".format(symbol.name)
+        if symbol.is_constant:
+            result += " = {0}".format(symbol.constant_value)
+        result += "\n"
+        return result
+
+    def node(self, node):
+        '''Catch any unsupported nodes, output their class names and continue
+        down the node hierarchy.
+
+        :param node: An unsupported PSyIR node.
+        :type node: subclass of :py:class:`psyclone.psyGen.Node`
+
+        :returns: The Fortran code as a string.
+        :rtype: str
+
+        '''
+        result = "{0}[ {1} start ]\n".format(self._nindent,
+                                             type(node).__name__)
+        self._depth += 1
+        for child in node.children:
+            result += self.visit(child)
+        self._depth -= 1
+        result += "{0}[ {1} end ]\n".format(self._nindent, type(node).__name__)
+        return result
```

Comment:
I understand this `node` visitor is useful for developing the back-end but ultimately this should raise an exception or remove the method and let the base class raise the appropriate exception right?

I am happy to leave this until the back-end is more mature, but maybe add a TODO or NOTE comment in this method.

Reply:
Good point. I decided to remove it as the reader should support all of the PSyIR.